fc_util/src/arg_parser.rs

```diff
@@ -0,0 +1,581 @@
+            format!("--{} <{}>", self.name, self.name)
+        } else {
+            format!("--{}", self.name)
+        }
+    }
+}
+
+impl<'a> PartialEq for ArgInfo<'a> {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name
+    }
+}
+
+impl<'a> Hash for ArgInfo<'a> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+    }
+}
+
+impl<'a> Borrow<str> for ArgInfo<'a> {
+    fn borrow(&self) -> &str {
+        self.name
+    }
+}
+
+#[derive(Debug, Default, PartialEq)]
+pub struct ArgParser<'a> {
+    args_matches: HashSet<ArgInfo<'a>>,
+    extra_args: Vec<String>,
+}
+
+impl<'a> ArgParser<'a> {
+    pub fn arg(mut self, arg_info: ArgInfo<'a>) -> Self {
+        self.args_matches.insert(arg_info);
+        self
+    }
+
+    pub fn value(&self, arg_name: &'static str) -> Option<String> {
+        self.args_matches.get(arg_name).unwrap().user_value.clone()
+    }
+
+    pub fn extra_args(self) -> Vec<String> {
+        self.extra_args
+    }
+
+    // Used for a general validation of the arguments provided by user.
+    fn is_valid(&self, argument_info: &ArgInfo<'a>, index: usize, args: &[String]) -> Result<()> {
+        // Check if the argument at `index` position was not provided more than once.
+        if self
+            .args_matches
+            .get(argument_info.name)
+            .unwrap()
+            .user_value
+            .is_some()
+        {
+            return Err(Error::DuplicatedArgument(argument_info.name.to_string()));
+        }
+        // If the argument at `index` position takes a value, check if there was provided
+        // one and only one value for that argument.
+        if argument_info.takes_value {
+            if let Some(arg_value) = args.get(index + 1) {
+                if arg_value.starts_with("--")
+                    && args.contains(&(format!("--{}", argument_info.name)))
+                {
+                    return Err(Error::MissingValue(argument_info.name.to_string()));
+                }
+            } else {
+                return Err(Error::MissingValue(argument_info.name.to_string()));
+            }
+        }
+        Ok(())
+    }
+
+    // Used for checking if user provided an argument at the "right" position relative to the
+    // previous argument.
+    fn next_arg(index: usize, args: &[String]) -> Result<()> {
+        if let Some(arg) = args.get(index) {
+            if !arg.starts_with("--") {
+                return Err(Error::UnexpectedArgument(arg.clone()));
+            }
+        }
+        Ok(())
+    }
+
+    pub fn parse(self) -> Result<Self> {
+        let mut args: Vec<String> = env::args().collect();
+        // Remove the first element of `args` as it should be the binary name.
+        args.remove(0);
+
+        self.get_matches(args)
+    }
+
+    // Populates the fields of ArgParser.
+    pub fn get_matches(mut self, args: Vec<String>) -> Result<Self> {
+        if args.contains(&"--help".to_string()) {
```

Comment:
I would define "--" as a constant. Also "help" and all the other hardcoded strings used in the code.